[cmd/opampsupervisor] Make configuration merge more customizable through special config files

douglascamata opened the pull request:
#### Description

This PR refactors the logic of the configuration loading and merging in the `cmd/opampsupervisor` package to allow for more customization, control and to have a more explicit ordering for users.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #39963.

<!--Describe what testing was performed and which tests were added.-->
#### Testing

Did some manual testing and ensured our current tests pass.

<!--Describe the documentation added.-->
#### Documentation

- Put very good notes on the changelog. 
- Update the specification docs.

<!--Please delete paragraphs that you did not use before submitting.-->

douglascamata committed:
Rename `loadRemoteConfig` to `loadLastRecvdRemoteConfig`

Rename v2

douglascamata committed:
Fix typo in logs

douglascamata committed:
Create variable only in the scope where it is used

douglascamata committed:
Implement magic config files for internal config

douglascamata committed:
Add changelog entry

douglascamata committed:
go fmt

douglascamata committed:
Avoid shadowing err

douglascamata committed:
Fix and improve specification docs on configuration

douglascamata committed:
Fix formatting

douglascamata committed:
Update example and remove repeated statement

atoulme commented:
Any chance to avoid the term magic? Would you please describe the intent, it looks like you have enough design going on a small RFC would help.

douglascamata commented:
> Any chance to avoid the term magic? 

The only other term that comes to my head and doesn't create even more confusion would be "special", which we could also use. Do you have any other suggestions, @atoulme? 

> Would you please describe the intent, it looks like you have enough design going on a small RFC would help.

We already have the initial intent in #39963, which was to be able to have local config files added into the configuration stack at the lowest priority level. The idea for the technical approach comes from [a suggestion in the same issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39963#issuecomment-2889014864). 

If really necessary for this I can try to write an RFC. @evan-bradley, by chance do you already have something written somewhere about [those future plans for writing all configs to disk and letting the Collector itself merge them all](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39963#issuecomment-2899079635)?

TylerHelmuth commented on cmd/opampsupervisor/specification/README.md:
```diff
@@ -220,19 +225,69 @@ telemetry:
-the configuration files specified are applied after the configuration from the OpAMP server.
-After the configuration files, arguments present in `.agent.args` are passed to the executable binary.
-The environmanet variables specified in `.agent.env` are set in the collector process environment.
+There are a few "magic" configuration files that can be used to completely
```
I vote we do not refer to these as `magic` configuration files. Something like `reserved` or `key` or `global`, `variable` or `special` as it feels more professional. I dont want users to think using these variables are something to be avoided.

TylerHelmuth commented on cmd/opampsupervisor/specification/README.md:
```diff
@@ -220,19 +225,69 @@ telemetry:
+The list below contains a few, but not all, examples of what can happen if a gven
+magic file is missing:
+
+- If the `$OPAMP_EXTENSION_CONFIG` is not specified, the Collector
+  will not be able to connect to the Supervisor, triggering a restart loop.
+- If the `$OWN_METRICS_CONFIG` is not specified, the Collector will not be able
+  to report its own metrics.
+- If the `$BUILTIN_CONFIG` is not specified, the Collector will not have the
+  expected identifying, non-identifying and resource attributes.
+- If the `$REMOTE_CONFIG` is not specified, the Collector will not be able to
+  apply the remote configuration received from the Supervisor.
```
Do we need to provide validation against the merged config that the things we expect to be present are present? I feel like we are opening up ourselves to a lot of troubleshooting issues from users holding this feature incorrectly

douglascamata replied:
@TylerHelmuth I thought about this a bit. One way we could avoid this problem is to automatically prepend any non-specified special file, so this way what we expect to be there will be there **unless** modified by later/higher-level configurations at the expense of having a bit of "hidden" behavior. I think this can be a good tradeoff. WDYT, @TylerHelmuth?

TylerHelmuth replied:
Ya always prepending one of our special configurations if it is not specified in the list by the user feels like a safer default.

evan-bradley replied:
I'm okay taking this on a case-by-case basis: for example, I'm not as worried about the own metrics config as the other ones.

However in general I think we should enable default anything that would cause the Supervisor to break in most cases (removing the OpAMP extension config) or introduce unclear behavior (the remote configuration capability is enabled, but the remote config merge order isn't set, so don't use remote config). For now I would be okay just always including all of the special configurations.

evan-bradley commented on cmd/opampsupervisor/specification/README.md:
```diff
@@ -220,19 +225,69 @@ telemetry:
+- `$BUILTIN_CONFIG`: built-in base configuration.
+- `$REMOTE_CONFIG`: remote configuration received by the Supervisor.
+
+If **none** of these are explicitly specified, they are automatically prepended to the
```
This would be a breaking change for the current behavior, but I would vote for having these **appended** to the list if they're unspecified:

1. The OpAMP extension configuration added by the Supervisor is essential for the Supervisor to work and disabling it should be something users have to explicitly opt into by putting the extension config before their custom config.
2. Ultimately, remote configuration is the "most specific" configuration, since it's been deliberately sent to a Collector by an operator. I think we should always respect this configuration unless the Collector has been specifically set up to override values that may be present in the remote configuration.

evan-bradley reviewed (commented):
Thanks for working on this @douglascamata. I've only looked at the specification so far since I want to get that right before focusing on the implementation.

douglascamata committed:
Address PR review feedback

douglascamata committed:
Update changelog

douglascamata committed:
Fix typo

douglascamata committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into supervisor-config-file-revamp

douglascamata committed:
Merge branch 'main' into supervisor-config-file-revamp

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Fix conflict resolution

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

evan-bradley reviewed (commented):
No real concerns about the approach. Thanks for taking this on.

douglascamata committed:
Apply suggestions from code review for comments

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

douglascamata commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -935,6 +934,63 @@ func (s *Supervisor) onOpampConnectionSettings(_ context.Context, settings *prot
+
+	// If there are config files AND none of them are the special ones, we
+	// add the special config files to the user provided ones.
+	if len(missingSpecialConfigFiles) == len(config.SpecialConfigFiles) {
+		newConfigFiles := []string{
+			string(config.SpecialConfigFileOwnMetrics),
+			string(config.SpecialConfigFileBuiltin),
+		}
+		newConfigFiles = append(newConfigFiles, s.config.Agent.ConfigFiles...)
+		newConfigFiles = append(newConfigFiles,
+			string(config.SpecialConfigFileOpAMPExtension),
+			string(config.SpecialConfigFileRemoteConfig),
+		)
+		s.config.Agent.ConfigFiles = newConfigFiles
+		return
+	}
```
Totally! Just wanted to make it slightly more performant/simpler, but it's not necessary and maybe not worth the duplication, so I wasn't very sure about it. Good to have a second opinion on it. I'll remove this and rely only on the logic below.

douglascamata commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -935,6 +934,63 @@ func (s *Supervisor) onOpampConnectionSettings(_ context.Context, settings *prot
+func (s *Supervisor) addSpecialConfigFiles() {
+	// - If no config file are provided, we add the special config files matching the
+	//   standard order of configuration merging.
+	if len(s.config.Agent.ConfigFiles) == 0 {
+		for _, file := range config.SpecialConfigFiles {
+			s.config.Agent.ConfigFiles = append(s.config.Agent.ConfigFiles, string(file))
+		}
+		return
+	}
```
In fact, as I started to do the change together with additional tests, I realized that to make the config validation function change the configs I would need to go around changing the method receivers to a reference instead of value. This makes me afraid of breaking other things unintentionally. So, if that's okay for you, I will keep only validation in the `Validate` method and keep the enforcements of defaults in this method of the Supervisor.

WDYT, @evan-bradley?

douglascamata replied:
BTW, I ended up removing manipulation of the config files from inside the Validate function (now only the `Supervisor.addSpecialConfigFiles` method does this): [`d3b4383` (#40295)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40295/commits/d3b4383fa4fb62b372217ede332852177a504558)

douglascamata committed:
Remove default config file enforcmeent from config validation

This logic will stay in the Supervisor. Config validation should not be modifying the config in any way.

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Simplify logic to add special config files

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Add test case for special config file validation

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Add a few test cases for `Supervisor.addSpecialConfigFiles()`

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into supervisor-config-file-revamp

douglascamata committed:
Retrigger build due to weird failure

Failure link: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/16021365184/job/45198938497?pr=40295

Exact failure message:

```
panic: failed to create SDK: binding address 127.0.0.1:38499 for Prometheus exporter: listen tcp 127.0.0.1:38499: bind: address already in use
```

douglascamata commented on cmd/opampsupervisor/supervisor/supervisor.go:
Tests for the logic in `Supervisor.addSpecialConfigFiles` added on [`ce6b08a` (#40295)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40295/commits/ce6b08aa15e945dc6b17c1b2d07db0b9221d6523)

douglascamata committed:
Add extra changelog entry for breaking change in agent config files behavior

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

evan-bradley commented:
>by chance do you already have something written somewhere about https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39963#issuecomment-2899079635?

@douglascamata Sorry I missed this before. I think this part can be implemented separately from the rest of this PR. It relates to the work in [this RFC](https://github.com/open-telemetry/opentelemetry-collector/pull/13256), though we may be able to just use the existing feature gate without any issues.

douglascamata commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -1017,34 +1091,55 @@ func (s *Supervisor) composeOpAMPExtensionConfig() []byte {
-		if err != nil {
-			s.telemetrySettings.Logger.Error("Could not read local config file", zap.Error(err))
+		// The special config files should always be valid yaml.
+		// If they are not we need a hard crash.
```
I can extract the logic into a separate function and call it in the Supervisor constructor. Let me try this.

evan-bradley replied:
Thanks for looking into this, that makes sense to me.

>In the case where we have a last received remote configuration saved, the Supervisor will exit as soon as it tries to rewrite its effective configuration file.

If I'm understanding you correctly, I think we should avoid shutting down the Supervisor for remote config in particular and instead report it as a bad config to the OpAMP server.

douglascamata replied:
Sorry, I wrote that wrong, @evan-bradley. 

If `composeMergedConfig` fails inside `processRemoteConfigMessage`, the Supervisor won't exit. You can see it here:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ff61b386f62d4e26f57d2517264686b1552b7252/cmd/opampsupervisor/supervisor/supervisor.go#L1754

douglascamata replied:
BTW I updated the comment to be more correct and less alarming.

douglascamata committed:
Fix typo (`OWN_METRICS` -> `OWN_TELEMETRY`)

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Add more tests for special config file handling

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata commented on cmd/opampsupervisor/supervisor/config/config.go:
```diff
@@ -218,9 +220,34 @@ func (a Agent) Validate() error {
+const (
+	SpecialConfigFileBuiltin        SpecialConfigFile = "$BUILTIN_CONFIG"
+	SpecialConfigFileOpAMPExtension SpecialConfigFile = "$OPAMP_EXTENSION_CONFIG"
+	SpecialConfigFileOwnMetrics     SpecialConfigFile = "$OWN_METRICS_CONFIG"
```
Fixed all occurrences in [`d239fb8` (#40295)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40295/commits/d239fb8919cd3cb8b7ea34cbc846185c327d5a45)

douglascamata committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into supervisor-config-file-revamp

douglascamata committed:
Fix comment in special config file loading

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

evan-bradley commented on cmd/opampsupervisor/specification/README.md:
```diff
@@ -220,19 +225,60 @@ telemetry:
+agent:
+  config_files:
+    - base_config.yaml
+    - $OWN_TELEMETRY_CONFIG
+    - $BUILTIN_CONFIG
```
Sorry to add this so late, but I think we should merge these files now that we're exposing them to users: looking at them, they both configure the Collector's telemetry, and I think we should just put them together if users need to understand what each is.

We could do this in a pre-req PR or in this PR, I'm fine either way.

evan-bradley replied:
The urgency for doing this as a prerequisite is that it will be a breaking change if we expose these options then change them, so we should just get it in now.

douglascamata replied:
That's a good idea. Let me see how big would the change be. If it's small enough I will include it here.

evan-bradley replied:
Whoops, sorry, I probably should have specified: could you call the resulting file `OWN_TELEMETRY_CONFIG`? I think that's the most accurate description of it right now.

douglascamata committed:
Fix changelog for breaking change

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Rename changelog file

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Merge `OWN_TELEMETRY_CONFIG` into `BUILTIN_CONFIG`

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata commented on .chloggen/supervisor-special-config-append.yaml:
```diff
@@ -0,0 +1,38 @@
+component: cmd/opampsupervisor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Any missing special configuration file is automatically added to the agent config list
```
Changelog fixed at [`a40349d` (#40295)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40295/commits/a40349dba957224e98accf0f88ecd5e9ddef170b)

douglascamata committed:
Rename built-in special config file

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Renames for consistency (`extra config` -> `extra telemetry config`, `own metrics` -> `own telemetry` )

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata committed:
Delete old `extraconfig.yaml` file

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

evan-bradley commented on cmd/opampsupervisor/supervisor/supervisor.go:
```diff
@@ -1022,34 +1091,57 @@ func (s *Supervisor) composeOpAMPExtensionConfig() []byte {
+type configComposer func() []byte
+
+func (s *Supervisor) composeAgentConfigFiles(incomingConfig *protobufs.AgentRemoteConfig) ([]byte, error) {
+	konf := koanf.New("::")
```
Could we call this `conf` or similar? `konf` surprised me at first when reading it because it reads like a misspelling of `koanf`.

evan-bradley commented on cmd/opampsupervisor/e2e_test.go:
```diff
@@ -1261,8 +1261,8 @@ func createBadCollectorConf(t *testing.T) (*bytes.Buffer, []byte) {
 
-func createHealthCheckCollectorConf(t *testing.T) (cfg *bytes.Buffer, hash []byte, remotePort int) {
-	colCfgTpl, err := os.ReadFile(path.Join("testdata", "collector", "healthcheck_config.yaml"))
+func createHealthCheckCollectorConf(t *testing.T, nopPipeline bool) (cfg *bytes.Buffer, hash []byte, remotePort int) {
```
Just so I understand, what's the change in behavior that required this change?

evan-bradley replied:
Thanks, I think I see the issue now. This probably warrants a second look at some point, but if the tests are passing I think we're okay for the moment.

evan-bradley reviewed (commented):
More or less looks good to me, just two questions.

douglascamata committed:
Rename `konf` to `conf`

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

evan-bradley reviewed (approved):
Thanks @douglascamata!

If you can fix the merge conflict, this looks good to me.

douglascamata committed:
Merge branch 'main' of github.com:open-telemetry/opentelemetry-collector-contrib into supervisor-config-file-revamp

Signed-off-by: Douglas Camata <159076+douglascamata@users.noreply.github.com>

douglascamata commented:
@evan-bradley conflict fixed. Thanks for bearing with me all this time. 🥂
